File: gapic/src/service/com/google/gapid/service/atom/Observation.java
/*
 * Copyright (C) 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.gapid.service.atom;

import com.google.gapid.rpclib.binary.BinaryClass;
import com.google.gapid.rpclib.binary.BinaryID;
import com.google.gapid.rpclib.binary.BinaryObject;
import com.google.gapid.rpclib.binary.Decoder;
import com.google.gapid.rpclib.binary.Encoder;
import com.google.gapid.rpclib.binary.Namespace;
import com.google.gapid.rpclib.schema.Array;
import com.google.gapid.rpclib.schema.Entity;
import com.google.gapid.rpclib.schema.Field;
import com.google.gapid.rpclib.schema.Method;
import com.google.gapid.rpclib.schema.Primitive;
import com.google.gapid.rpclib.schema.Struct;
import com.google.gapid.service.memory.MemoryRange;

import java.io.IOException;

public final class Observation implements BinaryObject {

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Observation that = (Observation)o;
    if (myRange != null ? !myRange.equals(that.myRange) : that.myRange != null) return false;
    if (myID != null ? !myID.equals(that.myID) : that.myID != null) return false;
    return true;
  }

  @Override
  public int hashCode() {
    int result = myRange != null ? myRange.hashCode() : 0;
    result = 31 * result + (myID != null ? myID.hashCode() : 0);
    return result;
  }

  @Override
  public String toString() {
    return "Observation{ID=" + myID + ", range=" + myRange + '}';
  }

  //<<<Start:Java.ClassBody:1>>>
  private MemoryRange myRange;
  private BinaryID myID;

  // Constructs a default-initialized {@link Observation}.
  public Observation() {}


  public MemoryRange getRange() {
    return myRange;
  }

  public Observation setRange(MemoryRange v) {
    myRange = v;
    return this;
  }

  public BinaryID getID() {
    return myID;
  }

  public Observation setID(BinaryID v) {
    myID = v;
    return this;
  }

  @Override
  public BinaryClass klass() { return Klass.INSTANCE; }


  private static final Entity ENTITY = new Entity("atom", "Observation", "", "");

  static {
    ENTITY.setFields(new Field[]{
      new Field("Range", new Struct(MemoryRange.Klass.INSTANCE.entity())),
      new Field("ID", new Array("id.ID", new Primitive("byte", Method.Uint8), 20)),
    });
    Namespace.register(Klass.INSTANCE);
  }
  public static void register() {}
  //<<<End:Java.ClassBody:1>>>
  public enum Klass implements BinaryClass {
    //<<<Start:Java.KlassBody:2>>>
    INSTANCE;

    @Override
    public Entity entity() { return ENTITY; }

    @Override
    public BinaryObject create() { return new Observation(); }

    @Override
    public void encode(Encoder e, BinaryObject obj) throws IOException {
      Observation o = (Observation)obj;
      e.value(o.myRange);
      o.myID.write(e);

    }

    @Override
    public void decode(Decoder d, BinaryObject obj) throws IOException {
      Observation o = (Observation)obj;
      o.myRange = new MemoryRange();
      d.value(o.myRange);
      o.myID = new BinaryID(d);
    }
    //<<<End:Java.KlassBody:2>>>
  }
}

File: core/gapil/snippets/observation.go
// Copyright (C) 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package snippets

import (
	"fmt"

	"github.com/google/gapid/framework/binary"
)

// observation is a snippet used to annotate a memory observation
type Observation ObservationType

func (Observation) isSnippet()      {}
func (Observation) finder() findFun { return findObservations }
func (Observation) IsEmpty() bool   { return false }

func (o Observation) String() string {
	return fmt.Sprintf("Observation(%s)", ObservationType(o))
}

// observations, is a collection of observations which is binary codable.
type Observations struct {
	binary.Generate
	path         Pathway
	observations []ObservationType
}

var _ KindredSnippets = &Observations{}

// findObservations can be used to build a observations collection which is a
// binary.Object which contains only observation snippets. Snippets of other
// types are return as an addition return value.
func findObservations(path Pathway, snippets Snippets) (KindredSnippets, Snippets) {
	observations := &Observations{path: path}
	remains := Snippets{}
	for _, s := range snippets {
		if l, ok := s.(Observation); ok {
			observations.observations = append(observations.observations, ObservationType(l))
		} else {
			remains = append(remains, s)
		}
	}
	return observations, remains
}
